avogadro/qtplugins/selectiontool/selectiontool.cpp

```diff
@@ -72,97 +76,92 @@ QWidget* SelectionTool::toolWidget() const
 {
-  if (e->button() != Qt::LeftButton || !m_renderer)
+  if (e->button() != Qt::LeftButton || !m_renderer) {
+    m_initSelectionBox = false;
     return nullptr;
+  }
 
   m_drawSelectionBox = false;
+  m_initSelectionBox = true;
   m_start = Vector2(e->pos().x(), e->pos().y());
   m_end = m_start;
-  Identifier hit = m_renderer->hit(e->pos().x(), e->pos().y());
-
-  // If an atom is clicked, accept the event, but don't add it to the atom list
-  // until the button is released (this way the user can cancel the click by
-  // moving off the atom, and the click won't get passed to the default tool).
-  if (hit.type == Rendering::AtomType)
-    e->accept();
-
-  // Remove the global accept to prevent the rectangle selection area from being
-  // rendered as the selection code is not yet in place for the scene.
   e->accept();
-
   return nullptr;
 }
 
 QUndoCommand* SelectionTool::mouseReleaseEvent(QMouseEvent* e)
 {
   // If the click is released on an atom, add it to the list
-  if (e->button() != Qt::LeftButton || !m_renderer)
+  if (e->button() != Qt::LeftButton || !m_renderer || m_doubleClick) {
+    m_doubleClick = false;
     return nullptr;
-
+  }
+  shouldClean(e);
   // Assess whether the selection box is big enough to use, or a mis-click.
-  bool bigEnough = false;
   m_end = Vector2(e->pos().x(), e->pos().y());
-  if (fabs(m_start.x() - m_end.x()) > 2 && fabs(m_start.y() - m_end.y()) > 2)
-    bigEnough = true;
+  Vector2f start(m_start.x() < m_end.x() ? m_start.x() : m_end.x(),
+                 m_start.y() < m_end.y() ? m_start.y() : m_end.y());
+  Vector2f end(m_start.x() > m_end.x() ? m_start.x() : m_end.x(),
+               m_start.y() > m_end.y() ? m_start.y() : m_end.y());
+  bool bigEnough =
+    fabs(start.x() - end.x()) > 2 && fabs(start.y() - end.y()) > 2;
 
   if (m_drawSelectionBox && bigEnough) {
-    auto hits =
-      m_renderer->hits(m_start.x(), m_start.y(), m_end.x(), m_end.y());
-
-    // Toggle the selection if the Ctrl modifier is pressed.
-    if (e->modifiers() & Qt::ControlModifier) {
-      for (auto it = hits.begin(); it != hits.end(); ++it) {
-        toggleAtom(*it);
-      }
-    } else {
-      // If the shift modifier is not pressed clear the previous selection.
-      if (!(e->modifiers() & Qt::ShiftModifier)) {
-        clearAtoms();
-      }
-      for (auto it = hits.begin(); it != hits.end(); ++it) {
-        addAtom(*it);
+    m_initSelectionBox = false;
+    auto hits = m_renderer->hits(start.x(), start.y(), end.x(), end.y());
+    for (auto it = hits.begin(); it != hits.end(); ++it) {
+      if (it->type == Rendering::AtomType) {
+        selectAtom(e, it->index);
       }
     }
   } else {
     // Single click
     m_start = Vector2(e->pos().x(), e->pos().y());
     m_end = m_start;
     Identifier hit = m_renderer->hit(e->pos().x(), e->pos().y());
-
     // Now add the atom on release.
     if (hit.type == Rendering::AtomType) {
-      toggleAtom(hit);
+      toggleAtom(hit.index);
```

Comment:
By moving the `shouldClean()` this should work for single atom toggle-on / toggle-off

Reply:
Yes. Funny thing is that now double click followed by ctrl click does a complementary selection.